### PR TITLE
Adding scoped top-N CMS monoid

### LINKED ISSUE #468: Scoped TopN CMS
Conceptually, we want behavior somewhat like `Map[K, TopNCMS[V]]`.  For each key, we separately store an estimate of the N most frequent values.  To save space we can instead use a single `TopCMS[(K, V)]`.  The meat of this would be a specialized `HeavyHittersLogic[(K, V)]` that only does evictions per key.

### PR DESCRIPTION
**Pull request by jnievelt**
Closes #468

A bit of refactoring out duplicate code.  Leaving in `TopPctCMSAggregator` and `TopNCMSAggregator` so this should be binary compatible.

### CODE REVIEW

**Comment by johnynek** [comment]
can you merge with develop to get the sbt fix?

**Comment by ianoc** [comment]
Sorry my bad, git foo on cmd line broke stuff and closed all of these

File: algebird-core/src/main/scala/com/twitter/algebird/CountMinSketch.scala [hunk 1]
```diff
@@ -780,7 +821,7 @@ case class TopNLogic[K](heavyHittersN: Int) extends HeavyHittersLogic[K] {
 
   require(heavyHittersN > 0, "heavyHittersN must be > 0")
 
-  override def purgeHeavyHitters(cms: CMS[K])(hitters: HeavyHitters[K]): HeavyHitters[K] = {
+  def purgeHeavyHitters(cms: CMS[K])(hitters: HeavyHitters[K]): HeavyHitters[K] = {
```
Comment (johnynek) [inline algebird-core/src/main/scala/com/twitter/algebird/CountMinSketch.scala]: why did you remove the override here?

File: algebird-core/src/main/scala/com/twitter/algebird/CountMinSketch.scala [hunk 2]
```diff
@@ -758,7 +799,7 @@ case class TopPctLogic[K](heavyHittersPct: Double) extends HeavyHittersLogic[K]
 
   require(0 < heavyHittersPct && heavyHittersPct < 1, "heavyHittersPct must lie in (0, 1)")
 
-  override def purgeHeavyHitters(cms: CMS[K])(hitters: HeavyHitters[K]): HeavyHitters[K] = {
+  def purgeHeavyHitters(cms: CMS[K])(hitters: HeavyHitters[K]): HeavyHitters[K] = {
```
Comment (johnynek) [inline algebird-core/src/main/scala/com/twitter/algebird/CountMinSketch.scala]: why no override? isn't this from HeavyHittersLogic?

File: algebird-core/src/main/scala/com/twitter/algebird/CountMinSketch.scala [hunk 3]
```diff
@@ -1045,14 +1017,83 @@ object TopNCMS {
 /**
  * An Aggregator for [[TopNCMS]].  Can be created using [[TopNCMS.aggregator]].
  */
-case class TopNCMSAggregator[K](cmsMonoid: TopNCMSMonoid[K])
-  extends MonoidAggregator[K, TopCMS[K], TopCMS[K]] {
+case class TopNCMSAggregator[K](cmsMonoid: TopNCMSMonoid[K]) extends TopCMSAggregator(cmsMonoid)
 
-  val monoid = cmsMonoid
+case class ScopedTopNLogic[K1, K2](heavyHittersN: Int) extends HeavyHittersLogic[(K1, K2)] {
```
Comment (johnynek) [inline algebird-core/src/main/scala/com/twitter/algebird/CountMinSketch.scala]: can we add a comment to this logic about what is going on? I think for each `K1`, we keep the top `heavyHittersN`.

**Comment by johnynek** [comment]
just a few minor comments, then let's merge.

File: algebird-core/src/main/scala/com/twitter/algebird/CountMinSketch.scala (hunk 1 above)
Comment (jnievelt) [inline algebird-core/src/main/scala/com/twitter/algebird/CountMinSketch.scala]: Eh I kinda love-hate the override keyword.  It doesn't require that the base class/trait actually implements the method (i.e., so that we know if super is interesting).  Then again, plain def doesn't require that the base class/trait declares the method either.  I'll put it back.
